Documentation/mds.bnf

```diff
@@ -0,0 +1,74 @@
+;;
+;; BNF formalization of MDS file grammar.
+;;
+
+;;
+;; This specification takes some liberties to reduce absurd verbosity.
+;; Care must be taken to ensure the extensions are explained in clear terms
+;; accompanied by proofs if necessary. Given the 'classes', identifiers
+;; between angle-brackets '<' and '>', as sets, set theory operations
+;; will be used to describe other classes when convenient. To make things
+;; clearer the regular equality sign will be used when the right side of a
+;; class definition uses set operations and their notation. General use
+;; of `where` will also be used for the addition of lexically scoped
+;; classes when convenient.
+;;
+
+<whitespace> ::= " "
+<ws> ::= <whitespace>
+
+<optional-whitespace> ::= " " | ""
+<opt-ws> ::= <optional-whitespace>
+
+;; <newline> is used to stick each component/record of the format together.
+;; As such, it is illegal inside <string>
+<newline> ::= \n | \r\n | <whitespace> <newline>
+<nl> ::= <newline>
+
+;; Any characters that are not valid for <newline>.
+<line> ::= <character'> | <character'> <string>
+  where
+    <character'> = <newline> △ <character>
+
+    ;; Do we need all of utf8?
+    ;; I tend to prefer generality, but maybe this is too much.
+    <character> ::= <set of all valid utf8>
```

Comment:
The main deviation I have made is here.  Both `<newline> △ <character>` and `<set of all valid utf8>` (I should add `codepoints` to that text) are shortcuts, but otherwise we need to define our `<string>` class by including EVERY character we want to be valid in this file.  Even with just lower and uppercase ascii that will add clutter, but if we want to make sure that Asian, Arabic, Jewish, etc... authors can add their real given name to MDS files, then the grammar should include more than ascii.  Defining such a class without shortcuts would mean a seriously long class definition where we would need to include all of the characters (many have to be represented as strings, since they are multibyte).

NB. Multibyte strings don't break the C or ascii rules, so allowing all utf8 codepoints doesn't require any changes in your code (probably).

Reply:
Given the massive class definitions that would be in the file without the Set based shortcuts I think it would be harder to implement parsers.  At least, to make it easier we would need comments explaining what the giant right side content is, so we would essentially have some ad hoc implementation of these shortcuts as informal comments.  Anyway, I just thought this perspective would help us think about it.  If you think the current approach is fine, then I guess we have no issue.  I will work on a formalization for the MDM files in whatever free time I get this week ;)

Reply:
Let's keep it that way then. No need to be 100% BNF compliant.

I tried to explore other ways this evening, it's just not worth it. Your current implementation is good!

Reply:
In case we want anything from the established extended BNFs https://en.m.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_form